core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java

```diff
@@ -79,9 +82,29 @@ public JoinHashSupplier(
             positionLinksFactoryBuilder = ArrayPositionLinks.builder(addresses.size());
         }
 
-        this.pages = channelsToPages(channels);
+        long additionalPagesRetainedSizeInBytes = 0;
+        // Currently PageIndex is retained by the HashBuilderOperator after JoinHash is created.
```

Comment:
Is or is-not. If it is retained then why do we need to account for this memory once again?

Reply:
This is rather convoluted. `HashBuilderOperator` retains `PageIndex`. When `PageIndex` is being built memory usage is set to `PageIndex#getEstimatedSize`. After `LookupSource` (`JoinHash`) is created memory usage of `HashBuilderOperator` is set to `JoinHash#getInMemorySizeInBytes` while a reference to `PageIndex` is still retained. I think the assumption is that the `JoinHash` retains same data what `PageIndex` does, and it had been (almost) true up until the introduction of `positionCounts`. However the overhead of pages created for `JoinHash` have never been accounted, that's also what I'm trying to accommodate here.

It feels like ideally it has to be refactored:

- Once `JoinHash` is created `PageIndex` has to be released. This is problematic with spilling, but should be doable in a version that doesn't support spilling
- `JoinHash` ideally has to be refactored to avoid this duality of having `List<Page>` in addition to `List<List<Block>>`. However it is not a straightforward change.